blog/2022/04/simplechains.md

```diff
@@ -0,0 +1,489 @@
+     title="Which Micro-optimizations matter for BLAS1?",
+     ylabel = "Time (s)", xlabel = "N",
+     yticks = 10.0 .^ (-8:0.5:2),)
+plot!(ns,noalloc,label=".=")
+plot!(ns, noalloc, label="mybroadcastavx!")
+plot!(ns,allocgpu,label="= gpu")
+plot!(ns,noallocgpu,label=".= gpu")
+savefig("microopts_blas1.png")
+```
+
+![](https://user-images.githubusercontent.com/1814174/162625049-b26fd0fd-271a-4c73-a44d-f40b36b18136.png)
+
+This already highly motivates a project focused on the performance for this case, but assumptions (3) and (4) point us to additionally look at the implementation of the backpropogation. The [trade-off between different machine learning library approaches to automatic differentiation has already been discussed at length](https://www.stochasticlifestyle.com/engineering-trade-offs-in-automatic-differentiation-from-tensorflow-and-pytorch-to-jax-and-julia/), but what the general discussions can miss is the extra opportunities afforded when really specializing on a domain. Take for example the use-case inside of neural ordinary differential equations (neural ODEs) and ODE adjoints. As mentioned above, in this use case the backwards pass is applied immediately after the forward pass. Thus while [a handwritten adjoint to a neural network layer](https://github.com/SciML/DiffEqFlux.jl/blob/v1.8.1/src/fast_layers.jl#L38-L56) can look like:
+
+```julia
+ZygoteRules.@adjoint function (f::FastDense)(x,p)
+  W = ...
+  b = ...
+  r = W*x .+ b
+  y = f.σ.(r)
+  function FastDense_adjoint(v)
+    σbar = ForwardDiff.derivative.(f.σ,r)
+    zbar = v .* σbar
+    Wbar = zbar * x'
+    bbar = zbar
+    xbar = W' * zbar
+    nothing,xbar,vcat(vec(Wbar),bbar)
+  end
+  y,FastDense_adjoint
+end
+```
+
+for `sigma.(W*x .+ b)` to calculate `J'v`, you can greatly optimize this if you know that the backwards pass will immediately preceed the forward pass. Specifically, there's no need to generate closures to store values since there no indeterminate future where the gradient might be needed, instead you can immediately proceed to calculate it. And, if you're only applying it to a vector of known size `v`, then this operation can be done non-allocating by mutating a cache vector. Lastly, if we know we're only going to use the derivative w.r.t. `x` (`xbar`), then we can eliminate many calculations. Look at the simplified version:
+
+```julia
+r = W*x .+ b
+y = σ.(r)
+σbar = derivative.(σ,r)
+zbar = v .* σbar
+Wbar = zbar * x'
+bbar = zbar
+xbar = W' * zbar   
+```
+
+and now cached:
+
+```julia
+mul!(cache,W,x)
+cache .= σ.(cache .+ b)
+cache .= derivative.(σ,cache)
+cache .= v .* cache
+mul!(xbar,W',cache)
+```
+
+or in other words, we can write this as a mutating operation with a single cache vector: `vjp!(xbar,W,b,σ,v,cache)`. All of the overheads of any automatic differentiation or mutations? Gone.
+
+Of course, building this up for anything other than the simplest case takes a much larger effort. In comes SimpleChains.jl.
```

Comment:
Should I implement `vjp!` before the SimpleChains?

The example that follows of course is just MNIST, an extremely "small" model, in that the training data takes only around 0.175 GiB of RAM, and the number of parameters is small compared to the absurdly huge numbers we've been hearing thrown around recently for networks like GPT3, but at 44_216 parameters, it's still absurdly massive compared to things like statistical models.

Reply:
It would be nice to specialize it in DiffEqSensitivity.jl. I guess we can wait and see what is and isn't optimized by the compiler, though to really know the effect it would take an implementation of a `vjp!` function